Web Inspector: Add Order Number to display: grid and grid-lanes

stwrt opened the pull request:
#### 10ae7dbc90309fc2e09c0cd83fdc7590481a322e
<pre>
Web Inspector: Add Order Number to display: grid and grid-lanes
<a href="https://bugs.webkit.org/show_bug.cgi?id=304287">https://bugs.webkit.org/show_bug.cgi?id=304287</a>
<a href="https://rdar.apple.com/problem/166648769">rdar://problem/166648769</a>

Reviewed by BJ Burg and Devin Rousso.

Add an &quot;Order Numbers&quot; option to CSS Grid overlays in Web Inspector, matching
the existing functionality for Flexbox overlays.

The implementation supports both regular CSS Grid layouts and CSS Masonry layouts.
For regular grids, item bounds are computed from grid area lines. For masonry
layouts, the item&apos;s frame rect is used directly since grid areas may not have
definite positions in the masonry axis.

* LayoutTests/inspector/dom/showFlexOverlay-expected.txt:
* LayoutTests/inspector/dom/showFlexOverlay.html:
* LayoutTests/inspector/dom/showGridOverlay-expected.txt:
* LayoutTests/inspector/dom/showGridOverlay.html:
* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::clearGridOverlayForNode):
(WebCore::InspectorOverlay::clearFlexOverlayForNode):
(WebCore::InspectorOverlay::buildGridOverlay):
(WebCore::InspectorOverlay::buildFlexOverlay):
* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.buildHighlightConfigs):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.set layoutFlags):
(WI.DOMNode.prototype.showLayoutOverlay):
(WI.DOMNode.prototype.hideLayoutOverlay):
* Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
(WI.LayoutDetailsSidebarPanel.prototype.initialLayout):

Canonical link: <a href="https://commits.webkit.org/304645@main">https://commits.webkit.org/304645@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7099f2d4cf9210b9fef2f5517c5570269807363e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143849 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138011 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8342 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104086 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4a34fb3b-b719-4c59-9163-5b9b9eeefa20) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139087 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6661 "Found 1 new test failure: fast/html/a-tooltip-null.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122000 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84925 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6334 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3992 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4445 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128099 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146596 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134625 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8181 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112443 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8197 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6874 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112777 "Found 1 new API test failure: WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/title-change (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28629 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6249 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118301 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8229 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36362 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167405 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7945 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71788 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43686 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8168 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8021 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1866,94 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
     }
 
+    if (gridOverlay.config.showOrderNumbers) {
+        Vector<RenderBox*> gridItemsInGridOrder;
```
These should probably be CheckedPtrs.

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -1234,6 +1234,8 @@ localizedStrings["Option-click to show unused CSS variables from all rules @ Sty
 localizedStrings["Options"] = "Options";
 /* Label for option to toggle the order numbers setting for CSS flex overlays */
 localizedStrings["Order Numbers @ Layout Panel Overlay Options"] = "Order Numbers";
```
NIT: i wonder if we should change this here (and where it's defined) to be clearer
```diff
-/* Label for option to toggle the order numbers setting for CSS flex overlays */
-localizedStrings["Order Numbers @ Layout Panel Overlay Options"] = "Order Numbers";
+/* Label for option to toggle the order numbers setting for CSS grid overlays */
+localizedStrings["Order Numbers @ Layout Panel Flex Overlay Options"] = "Order Numbers";
```

alternatively, you could remove the mention of `flex` (and `grid`) altogether so that both uses can share the same localized string

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -1234,6 +1234,8 @@ localizedStrings["Option-click to show unused CSS variables from all rules @ Sty
 /* Label for option to toggle the order numbers setting for CSS flex overlays */
 localizedStrings["Order Numbers @ Layout Panel Overlay Options"] = "Order Numbers";
+/* Label for option to toggle the order numbers setting for CSS grid overlays */
+localizedStrings["Order Numbers @ Grid Layout Panel Overlay Options"] = "Order Numbers";
```
NIT: i'd say this instead
```diff
-localizedStrings["Order Numbers @ Grid Layout Panel Overlay Options"] = "Order Numbers";
+localizedStrings["Order Numbers @ Layout Panel Grid Overlay Options"] = "Order Numbers";
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
```diff
@@ -74,6 +74,7 @@ WI.LayoutDetailsSidebarPanel = class LayoutDetailsSidebarPanel extends WI.DOMDet
         gridSettingsGroup.addSetting(WI.settings.gridOverlayShowLineNames, WI.UIString("Line Names", "Line names @ Layout Panel Overlay Options", "Label for option to toggle the line names setting for CSS grid overlays"));
         gridSettingsGroup.addSetting(WI.settings.gridOverlayShowAreaNames, WI.UIString("Area Names", "Area names @ Layout Panel Overlay Options", "Label for option to toggle the area names setting for CSS grid overlays"));
         gridSettingsGroup.addSetting(WI.settings.gridOverlayShowExtendedGridLines, WI.UIString("Extended Grid Lines", "Show extended lines @ Layout Panel Overlay Options", "Label for option to toggle the extended lines setting for CSS grid overlays"));
+        gridSettingsGroup.addSetting(WI.settings.gridOverlayShowOrderNumbers, WI.UIString("Order Numbers", "Order Numbers @ Grid Layout Panel Overlay Options", "Label for option to toggle the order numbers setting for CSS grid overlays"));
```
this should only be shown when inspecting something that supports it (i.e. not older backends)
```js
// COMPATIBILITY (macOS X.Y, iOS X.Y): `DOM.GridOverlayConfig.showOrderNumbers` did not exist yet.
// Since support can't be tested directly, check for if the `Page.navigate` command has been removed.
// FIXME: Use explicit version checking once https://webkit.org/b/148680 is fixed.
if (!InspectorBackend.hasCommand("Page.navigate"))
    gridSettingsGroup.addSetting(WI.settings.gridOverlayShowOrderNumbers, WI.UIString("Order Numbers", "Order Numbers @ Layout Panel Grid Overlay Options", "Label for option to toggle the order numbers setting for CSS grid overlays"));
```

dcrousso reviewed (approved):
neat!

only one thing that needs fixing :)

rcaliman-apple commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1864,91 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            StringBuilder orderNumbers;
+
+            if (auto index = gridItemsInDOMOrder.find(gridItem); index != notFound)
+                orderNumbers.append("Item #"_s, index + 1);
```
How does this string get localized?

stwrt replied:
It currently doesn't. Uploading a new patch to address it.

burg commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1864,91 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                auto absoluteRect = FloatRect(gridItem->absoluteBoundingBoxRect(true));
+                auto margins = gridItem->marginBox();
+                absoluteRect.expand(FloatBoxExtent(margins.top(), margins.right(), margins.bottom(), margins.left()));
+                // Convert to root view coordinates
```
Nit: this comment is redundant

burg commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1864,91 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            if (renderGrid.isMasonry()) {
+                // For masonry layouts, use absoluteBoundingBoxRect to get the visual position
+                // accounting for all scroll offsets and transforms including zoom.
+                auto absoluteRect = FloatRect(gridItem->absoluteBoundingBoxRect(true));
```
Nit: brace initializers if possible

burg commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1864,91 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            StringBuilder orderNumbers;
+
+            if (auto index = gridItemsInDOMOrder.find(gridItem); index != notFound)
+                orderNumbers.append("Item #"_s, index + 1);
```
Personally I would drop the #, it makes the overlay more busy.

Also, we probably need to localize this string somehow, since it is not a literal value like `order`, the expectation is that the label and placement of the digits would respect the user locale.

stwrt replied:
`#` is used by flex box too, so I would prefer we do not deviate.

stwrt replied:
<img width="922" height="248" alt="image" src="https://github.com/user-attachments/assets/fda23d37-5467-4ab4-84d9-a26b2f07c099" />

Flex box example

stwrt replied:
I would be happy to do a follow up patch where we change both then if we think that is better.

burg commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1866,6 +1864,91 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                orderNumbers.append("Item #"_s, index + 1);
+
+            if (auto order = gridItem->style().order(); !order.isZero() || hasCustomOrder)
+                orderNumbers.append(orderNumbers.isEmpty() ? ""_s : "\n"_s, "order: "_s, order.value);
```
Ditto re: localization

burg commented on Source/WebCore/inspector/InspectorOverlay.h:
```diff
@@ -163,6 +163,7 @@ class InspectorOverlay : public CanMakeWeakPtr<InspectorOverlay> {
             bool showExtendedGridLines;
             bool showTrackSizes;
             bool showAreaNames;
+            bool showOrderNumbers;
```
It's a little weird that we don't have default value initializers for this struct, but I guess we default it later.

burg reviewed (approved):
Nice!

stwrt committed:
Web Inspector: Add Order Number to display: grid and grid-lanes
https://bugs.webkit.org/show_bug.cgi?id=304287
rdar://problem/166648769

Reviewed by BJ Burg and Devin Rousso.

Add an "Order Numbers" option to CSS Grid overlays in Web Inspector, matching
the existing functionality for Flexbox overlays.

The implementation supports both regular CSS Grid layouts and CSS Masonry layouts.
For regular grids, item bounds are computed from grid area lines. For masonry
layouts, the item's frame rect is used directly since grid areas may not have
definite positions in the masonry axis.

* LayoutTests/inspector/dom/showFlexOverlay-expected.txt:
* LayoutTests/inspector/dom/showFlexOverlay.html:
* LayoutTests/inspector/dom/showGridOverlay-expected.txt:
* LayoutTests/inspector/dom/showGridOverlay.html:
* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::clearGridOverlayForNode):
(WebCore::InspectorOverlay::clearFlexOverlayForNode):
(WebCore::InspectorOverlay::buildGridOverlay):
(WebCore::InspectorOverlay::buildFlexOverlay):
* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.buildHighlightConfigs):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.set layoutFlags):
(WI.DOMNode.prototype.showLayoutOverlay):
(WI.DOMNode.prototype.hideLayoutOverlay):
* Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
(WI.LayoutDetailsSidebarPanel.prototype.initialLayout):

Canonical link: https://commits.webkit.org/304645@main
